src/OsuSharp/RateLimiter.cs

```diff
@@ -1,74 +1,65 @@
 ﻿using System;
 using System.Threading;
 using System.Threading.Tasks;
-using OsuSharp.Enums;
-using OsuSharp.Interfaces;
 
 namespace OsuSharp
 {
-    internal sealed class RateLimiter : IRateLimiter
+    internal sealed class RateLimiter
     {
-        private SemaphoreSlim Semaphore { get; }
-
-        private int Requests { get; set; }
-
-        private int MaxRequests { get; }
-
-        private TimeSpan TimeInterval { get; }
+        private RateLimiterConfiguration Configuration { get; }
 
-        private DateTime Time { get; set; }
+        private SemaphoreSlim Semaphore { get; }
 
-        private IOsuSharpLogger Logger { get; }
+        private int RequestCount { get; set; }
 
-        private bool ThrowsOnMaxRequests { get; }
+        private DateTimeOffset TimeReference { get; set; }
 
-        public RateLimiter(int maxRequests, TimeSpan timeInterval, bool throwsOnMaxRequests, IOsuSharpLogger logger)
+        internal RateLimiter(RateLimiterConfiguration rateLimiterConfiguration)
         {
-            MaxRequests = maxRequests;
-            TimeInterval = timeInterval;
-            ThrowsOnMaxRequests = throwsOnMaxRequests;
-            Logger = logger;
+            Configuration = rateLimiterConfiguration;
 
             Semaphore = new SemaphoreSlim(1);
 
-            Time = DateTime.Now + timeInterval;
+            RequestCount = 0;
+
+            TimeReference = DateTimeOffset.Now + Configuration.Interval;
         }
 
-        public async Task HandleAsync(CancellationToken cancellationToken = default)
+        internal async Task HandleAsync(CancellationToken token = default)
```

Comment:
Should prob use `CancellationToken.None` instead

Reply:
That's exactly what that is and you cannot set values that aren't compile-time constants as default parameter values.